cilium/CFP-23380-ordered-policy.md

```diff
@@ -0,0 +1,329 @@
+    rule2["prio 2, tcp, 80, id {*}, disabled"]
+    rule3["prio 3, tcp, 80, id {5}, disabled"]
+    root[[root]]
+    
+    root --> rule1
+    rule1 -. invalid .-> rule2 & rule3
+```
+
+#### Example 2b: Descendant wildcard
+
+Two rules:
+1. `(tcp, port *, id {5, 6})`
+2. `(tcp, port 80, id {*})`
+
+Rule 2 is higher precedence. For ID 5, rule 2 overrides rule 1, which is an error. However, rule 2's wildcard selector selects more identities, so it cannot be elided. Rather, we explicitly insert covering entry `(tcp, port 80, id {5, 6})`. Likewise, we flag rule 1 as "covered" by rule 2, so we know to manage incremental updates.
+
+```mermaid
+
+graph TB
+    rule1["prio 1, tcp, 0-65535, id {5, 6}"]
+    rule2["prio 2, tcp, 80, id {*}"]
+    rule1a["prio 1, tcp, 80, id {5, 6}"]
+    root[[root]]
+    
+    root --> rule1
+    rule1 --> rule2 & rule1a
+    rule2 -. covers .-> rule1
+```
+
+#### Merging mutual authentication
+
+Whether or not a connection requires mututal authentication is *also* managed by the network policy engine. However, authentication is not a "traffic verdict". (Traffic verdicts are `allow`, `deny`, and `proxy`). Rather, it is applied to all traffic, regardless of selectors.
+
+There are two options for resolving mutual authentication. The first is to push auth down to all mapstate entries as part of resolving conflicts. The other is to maintain a parallel overlay selector trie for authentcation, which is then merged when generating the BPF entries. In order to keep the merge algorithm simpler, an overlay trie is our current proposal.
+
+### Processing incremental updates
+
+In incremental update is when a complete policy newly selects, or no longer selects, one or more identities. These incremental updates must then be propagated down to the MapState and the BPF map updated accordingly.
+
+Incremental updates are aggregated by selectors and then inserted in to the MapState. The MapState is responsible for deconflicting overlapping selectors. This remains the same in the ordered version.
+
+The important change is that incremental updates must be inserted in to the map state *in priority order*. That is to say, higher-priority PerSelectorPolicies must apply their updates first. Then, the usual MapState deconfliction rules apply, and there is no risk of an incremental update diverging from policy intent.
+
+It is safe to insert incremental updates in this manner because identities and selectors are immutable. As a direct consequence, two selectors S1 and S2 might overlap, a new identity I1 can never change state from non-overlapping to overlapping.
+
+We will need to ensure that SelectorCache updates are processed in a transactional manner -- that is to say, an incremental update must be delivered to all affected selectors *before* distributing this to the MapState.
+
+#### Example 1: no wildcards
+
+Consider a set of rules:
+
+1. tcp 0-1023, ID (5,6)
+2. tcp 80, ID (7)
+3. tcp *, ID (8)
```

Comment:
Are the numbers at the start of the lines here representing priority? This should be more explicit to describe the meaning of the ordered list.